theseus/labs/lie/lie_tensor.py

```diff
@@ -143,6 +314,67 @@ def _no_jop(self, input0: TensorType) -> _JFnReturnType:
     jleft_act = _no_jop
     jleft_project = _no_jop
 
+    def retract(self, delta: TensorType) -> "LieTensor":
+        assert isinstance(delta, torch.Tensor)
+        return self.compose(self.exp(delta))
+
+    # ------------------------------------------------------
+    # Overloaded python and torch operators
+    # ------------------------------------------------------
+    def __add__(self, other: TensorType) -> "LieTensor":
+        if not isinstance(other, TangentTensor):
+            raise RuntimeError(
+                "Operator + is only supported for tensors of ltype=tgt. "
+                "If you intend to add the raw tensor data, please use group._t. "
+                "If you intend to retract, then cast your tensor to ltype=tgt, or "
+                "use group.retract(tensor)."
+            )
+        return self.retract(other._t)
+
+    def set_(self, tensor: "LieTensor"):  # type: ignore
+        if not isinstance(tensor, LieTensor):
+            raise RuntimeError(
+                "LieTensor.set_ is only supported for LieTensor arguments."
+            )
+        if not tensor.ltype == self.ltype:
+            raise RuntimeError(
+                f"Tried to set a tensor of type {self.ltype} with a "
+                f"tensor of type {tensor.ltype}"
+            )
+        super().set_(tensor)  # type: ignore
+
```

Comment:
A less esoteric but perhaps less safe alternative for `add_` could be to check the tensor shape and if it looks like (...,3, 4) (for SE3) this is interpreted as an euclidean gradient, something with shape `(..., 6)`interpret this as a Riemannian graadient, and everything else throws an error.